src/testing_tracer.rs

```diff
@@ -0,0 +1,191 @@
+use circuit_definitions::zk_evm::vm_state::ErrorFlags;
+use circuit_definitions::zk_evm::vm_state::PrimitiveValue;
+use zkevm_assembly::zkevm_opcode_defs::AddOpcode;
+use zkevm_assembly::zkevm_opcode_defs::NopOpcode;
+use zkevm_assembly::zkevm_opcode_defs::Opcode;
+use zkevm_assembly::zkevm_opcode_defs::PtrOpcode;
+use zkevm_assembly::zkevm_opcode_defs::RetOpcode;
+
+use crate::ethereum_types::U256;
+use crate::zk_evm::opcodes::DecodedOpcode;
+use crate::zk_evm::reference_impls::memory::SimpleMemory;
+use crate::zk_evm::tracing::*;
+
+use crate::asm_templates::EXCEPTION_PREFIX;
+use crate::asm_templates::PRINT_PREFIX;
+use crate::asm_templates::PRINT_PTR_PREFIX;
+use crate::asm_templates::PRINT_REG_PREFIX;
+
+/// Tracks prints and exceptions during VM execution cycles.
+#[derive(Debug, Clone)]
+pub struct TestingTracer {
+    /// the last uncatched exception message
+    pub exception_message: Option<String>,
+    /// true if next command should be interpreted as printable value and printed
+    expecting_register_value: bool,
+}
+
+/// TestingTracer interprets valid x values in `add x r0 r0` and `ptr.add x r0 r0` instructions as commands to execute.
+/// Commands have following structure: "PREFIX:arg"
+/// Allowed commands:
+/// "EXCEPTION_PREFIX:<text>" - save <text> in exception_message field
+/// "PRINT_PREFIX:<text>" - print <text> in the console
+/// "PRINT_REG_PREFIX:" - print raw "x" value of next command in the console
+/// "PRINT_PTR_PREFIX:" - print raw "x" pointer value of next command in the console (currently same result as previous command)
+impl TestingTracer {
+    pub fn new() -> Self {
+        Self {
+            exception_message: None,
+            expecting_register_value: false,
+        }
+    }
+
+    fn reset_exception(&mut self) {
+        self.exception_message = None;
+    }
+
+    fn set_exception_message(&mut self, message: &str) {
+        self.exception_message = Some(message.to_owned());
+    }
+
+    fn execute_print(&mut self, message: &str) {
+        println!("{}", message);
+    }
+
+    fn execute_print_from_register(&mut self, val: PrimitiveValue) {
+        assert!(
+            self.expecting_register_value,
+            "Unexpected print_from_register command"
+        );
+        self.expecting_register_value = false;
+        println!("{}", val.value);
+    }
+
+    fn parse_command_from_register(
+        &mut self,
+        val: PrimitiveValue,
+    ) -> (Option<String>, Option<String>) {
+        if val.value == U256::from(0) {
+            return (None, None);
+        }
+
+        let bytes: &mut [u8; 32] = &mut [0; 32];
+        val.value.to_big_endian(bytes);
+
+        if let Ok(message) = std::str::from_utf8(bytes) {
+            let message_trimmed = message.trim_matches(char::from(0));
```

Comment:
You can do:
```rust
        let mut bytes: [u8; 32] = [0; 32];
        val.value.to_big_endian(&mut bytes);

        if let Ok(message) = std::str::from_utf8(&bytes) {
```

this way 'bytes' is a 'simple' array - and then you pass either normal or mutable references.